src/windows/osd.c

```diff
@@ -100,6 +100,55 @@ int socketpair(int af, int type, int protocol, int socks[2])
 	return SOCKET_ERROR;
 }
 
+int fi_read_file(const char *dir, const char *file, char *buf, size_t size)
+{
+	char *path = 0;
+	int len, lendir, lenfile, pathlen;
+
+	HANDLE fd = INVALID_HANDLE_VALUE;
+	DWORD read;
+
+	len = -1;
+
+	lendir = lstrlenA(dir);
+	lenfile = lstrlenA(file);
+
+	pathlen = lendir + lenfile + 2; /* dir + '\' + file + '0' */
+
+	path = malloc(pathlen);
+	if (!path)
+		goto fn_fail;
+
+	lstrcpyA(path, dir);
+	if (lenfile) {
+		lstrcatA(path, "\\");
+		lstrcatA(path, file);
+	}
+
+	fd = CreateFileA(path, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, 0, OPEN_EXISTING, 0, 0);
+	if (fd == INVALID_HANDLE_VALUE)
+		goto fn_fail;
+
+	if (!ReadFile(fd, buf, (DWORD)size, &read, 0))
+		goto fn_fail;
+
+	len = (int)read;
+
+	if (len > 0 && buf[len - 1] == '\n')
+		buf[--len] = '\0';
+
+fn_exit:
+	if (fd != INVALID_HANDLE_VALUE)
+		CloseHandle(fd);
+	if (path)
+		free(path);
+	return len;
+
+fn_fail:
+	len = -1;
+	goto fn_exit;
+}
```

Comment:
I'm going to go ahead and merge this, but the preferred flow for error handing would be something like this:
```c
    ret = work1();
    if (ret != SUCCESS)
        return ret;
    ret = work2();
    if (ret != SUCCESS)
        goto err1;
    ret = work3();
    if (ret != SUCCESS)
        goto err2;
    blah_blah();
    return SUCCESS;
err2:
    undo_work2();
err1:
    undo_work1();
    return ret;